File: lectures/lecture-09-why-agents-declare-victory-too-early/exercises/exercise-02-layered-gate/starter/typescript/main.ts
// layered-gate exercise, TypeScript starter.
//
// The gate runs every declared check, groups the results into the three
// layers, and names the first failing layer in its verdict, but one naive
// decision remains (see SPEC.md "Starter state"): layers below a failing
// one are still executed and reported as if they meant something. Make
// the gate stop at the first failing layer, reporting later checks as
// not-reached, then run ../../verify.sh --stack=typescript until it
// exits 0.

import { existsSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const LAYERS = ["static", "tests", "system"];

interface ValueSide {
  path: string;
  key: string;
}

interface Check {
  id: string;
  layer: string;
  cost: number;
  kind: string;
  path?: string;
  prefix?: string;
  marker?: string;
  left?: ValueSide;
  right?: ValueSide;
}

interface Row {
  id: string;
  status: string;
  detail: string;
}

function fileExists(workspace: string, name: string): boolean {
  const path = join(workspace, name);
  return existsSync(path) && statSync(path).isFile();
}

function readLines(workspace: string, name: string): string[] {
  return readFileSync(join(workspace, name), "utf8").split(/\r?\n/);
}

function readKey(workspace: string, path: string, key: string): string | null {
  for (const line of readLines(workspace, path)) {
    if (line.startsWith(`${key}=`)) return line.slice(key.length + 1).trim();
  }
  return null;
}

function executeCheck(workspace: string, check: Check): [boolean, string] {
  if (check.kind === "file-exists") {
    const path = check.path as string;
    if (fileExists(workspace, path)) return [true, `${path} present`];
    return [false, `${path} missing`];
  }
  if (check.kind === "file-has-line") {
    const path = check.path as string;
    const prefix = check.prefix as string;
    if (!fileExists(workspace, path)) return [false, `${path} missing`];
    if (readLines(workspace, path).some((line) => line.startsWith(prefix))) {
      return [true, `${path} has a line starting with ${prefix}`];
    }
    return [false, `${path} has no line starting with ${prefix}`];
  }
  if (check.kind === "file-lacks-marker") {
    const path = check.path as string;
    const marker = check.marker as string;
    if (!fileExists(workspace, path)) return [false, `${path} missing`];
    if (readFileSync(join(workspace, path), "utf8").includes(marker)) {
      return [false, `${path} contains ${marker}`];
    }
    return [true, `${path} carries no ${marker} marker`];
  }
  if (check.kind === "values-agree") {
    const left = check.left as ValueSide;
    const right = check.right as ValueSide;
    for (const side of [left, right]) {
      if (!fileExists(workspace, side.path)) return [false, `${side.path} missing`];
      if (readKey(workspace, side.path, side.key) === null) {
        return [false, `${side.path} has no ${side.key}= line`];
      }
    }
    const leftValue = readKey(workspace, left.path, left.key);
    const rightValue = readKey(workspace, right.path, right.key);
    if (leftValue === rightValue) {
      return [
        true,
        `${left.path} ${left.key}=${leftValue} matches ${right.path} ${right.key}=${rightValue}`,
      ];
    }
    return [
      false,
      `${left.path} ${left.key}=${leftValue} but ${right.path} ${right.key}=${rightValue}`,
    ];
  }
  throw new Error(`unknown check kind: ${check.kind}`);
}

export function runLayers(workspace: string) {
  const config = JSON.parse(readFileSync(join(workspace, "checks.json"), "utf8")) as {
    checks: Check[];
  };
  const layers = [];
  let stoppedAt: string | null = null;
  for (const layer of LAYERS) {
    const declared = config.checks.filter((check) => check.layer === layer);
    // Naive draft: run every layer so the report is complete, and let the
    // verdict name where a stricter gate would have stopped. Exercise:
    // once a layer fails, later layers must not execute; report their
    // checks as not-reached, gated by the failing layer.
    const rows: Row[] = declared.map((check) => {
      const [passed, detail] = executeCheck(workspace, check);
      return { id: check.id, status: passed ? "pass" : "fail", detail };
    });
    const status = rows.every((row) => row.status === "pass") ? "passed" : "failed";
    if (status === "failed" && stoppedAt === null) stoppedAt = layer;
    layers.push({ layer, status, checks: rows });
  }
  return {
    workspace: workspace.replace(/\/+$/, "").split("/").pop() ?? workspace,
    layers,
    verdict: {
      stopped_at: stoppedAt,
      result: stoppedAt === null ? "done" : "not-done",
    },
  };
}

function main(argv: readonly string[]): number {
  const workspace = argv[2];
  if (argv.length !== 3 || !workspace) {
    console.error("usage: main.ts <workspace-dir>");
    return 2;
  }
  if (!existsSync(workspace) || !statSync(workspace).isDirectory() || !fileExists(workspace, "checks.json")) {
    console.error(`error: not a workspace (needs checks.json): ${workspace}`);
    return 2;
  }
  const report = runLayers(workspace);
  process.stdout.write(JSON.stringify(report, null, 2) + "\n");
  return report.verdict.result === "done" ? 0 : 1;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}

File: lectures/lecture-09-why-agents-declare-victory-too-early/exercises/exercise-02-layered-gate/starter/python/main.py
"""layered-gate exercise, Python starter.

The gate runs every declared check, groups the results into the three
layers, and names the first failing layer in its verdict, but one naive
decision remains (see SPEC.md "Starter state"): layers below a failing
one are still executed and reported as if they meant something. Make the
gate stop at the first failing layer, reporting later checks as
not-reached, then run ../../verify.sh --stack=python until it exits 0.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

LAYERS = ["static", "tests", "system"]


def read_key(path: Path, key: str) -> str | None:
    for line in path.read_text(encoding="utf-8").split("\n"):
        if line.startswith(f"{key}="):
            return line[len(key) + 1 :].strip()
    return None


def execute_check(workspace: Path, check: dict) -> tuple[bool, str]:
    kind = check["kind"]
    if kind == "file-exists":
        path = check["path"]
        if (workspace / path).is_file():
            return True, f"{path} present"
        return False, f"{path} missing"
    if kind == "file-has-line":
        path, prefix = check["path"], check["prefix"]
        file = workspace / path
        if not file.is_file():
            return False, f"{path} missing"
        if any(line.startswith(prefix) for line in file.read_text(encoding="utf-8").split("\n")):
            return True, f"{path} has a line starting with {prefix}"
        return False, f"{path} has no line starting with {prefix}"
    if kind == "file-lacks-marker":
        path, marker = check["path"], check["marker"]
        file = workspace / path
        if not file.is_file():
            return False, f"{path} missing"
        if marker in file.read_text(encoding="utf-8"):
            return False, f"{path} contains {marker}"
        return True, f"{path} carries no {marker} marker"
    if kind == "values-agree":
        left, right = check["left"], check["right"]
        for side in (left, right):
            if not (workspace / side["path"]).is_file():
                return False, f"{side['path']} missing"
            if read_key(workspace / side["path"], side["key"]) is None:
                return False, f"{side['path']} has no {side['key']}= line"
        left_value = read_key(workspace / left["path"], left["key"])
        right_value = read_key(workspace / right["path"], right["key"])
        if left_value == right_value:
            return True, (
                f"{left['path']} {left['key']}={left_value} matches "
                f"{right['path']} {right['key']}={right_value}"
            )
        return False, (
            f"{left['path']} {left['key']}={left_value} but "
            f"{right['path']} {right['key']}={right_value}"
        )
    raise ValueError(f"unknown check kind: {kind}")


def run_layers(workspace: Path) -> dict:
    config = json.loads((workspace / "checks.json").read_text(encoding="utf-8"))
    layers = []
    stopped_at = None
    for layer in LAYERS:
        declared = [check for check in config["checks"] if check["layer"] == layer]
        # Naive draft: run every layer so the report is complete, and let
        # the verdict name where a stricter gate would have stopped.
        # Exercise: once a layer fails, later layers must not execute;
        # report their checks as not-reached, gated by the failing layer.
        rows = []
        for check in declared:
            passed, detail = execute_check(workspace, check)
            status = "pass" if passed else "fail"
            rows.append({"id": check["id"], "status": status, "detail": detail})
        status = "passed" if all(row["status"] == "pass" for row in rows) else "failed"
        if status == "failed" and stopped_at is None:
            stopped_at = layer
        layers.append({"layer": layer, "status": status, "checks": rows})
    return {
        "workspace": workspace.name,
        "layers": layers,
        "verdict": {
            "stopped_at": stopped_at,
            "result": "done" if stopped_at is None else "not-done",
        },
    }


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <workspace-dir>", file=sys.stderr)
        return 2
    workspace = Path(argv[1])
    if not workspace.is_dir() or not (workspace / "checks.json").is_file():
        print(f"error: not a workspace (needs checks.json): {workspace}", file=sys.stderr)
        return 2
    report = run_layers(workspace)
    print(json.dumps(report, indent=2))
    return 0 if report["verdict"]["result"] == "done" else 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))
